### PR TITLE
EGP Enhancements and Fixes

### LINKED ISSUE #1251: Allow adding complete egp objects as table
While the current way of adding egp objects works and is simple to understand, it is cumbersome and fairly inefficent when adding complex objects.

```
EGP:egpTextLayout(Index, Text, Pos, Size)
EGP:egpFont(Index, Font, FontSize)
EGP:egpColor(Index, Color)
EGP:egpAlign(Index,1,1)
EGP:egpParent(Index, OtherIndex)
#55 ops, lots of avoidable (internal) calls
```
compared to something like
```
Object=table("type"="TextLayout", "pos"=Pos, "size"=Size, "font"=Font,
    "fontSize"=FontSize, "rgba"=Color, "alignment"=vec2(1,1), "parent"=OtherIndex)
EGP:egpAddObject(Index, Object)
```
(using different keys such as rgb/rgba to handle functions with multiple possible parameter choices, incase it should be neccessary at all)

If no type is set, it would merge the new data with the existing object, instead of overwriting it as a whole.

Other functions that would make sense in this context:
```
Object=EGP:egpGetObject(Index)

#Multiple Objects
EGP:egpAddObjects(table(Index1=Object1, Index2=Object2))

Objects=EGP:egpGetObjects(array(Index1,Index2))
AllObjects=EGP:egpGetObjects()
```

The function to add multiple objects would make sense as EGPs don't usually contain just a few objects, but dozends, with most of them created in bulk.  
It would also add a more general way to do batch operations compared to #967.

Reading back objects would be nice too, while only useful for a few purposes such as serialisation, it should be comparatively easy to just output a copy of the internal table.

### LINKED ISSUE #2190: EGP Boxes' parent origin is in center even after EGP:egpDrawTopLeft(1) call

### LINKED ISSUE #2349: Tone mapping messes with digital screen colors
image

### LINKED ISSUE #2546: [EGP] Lua error: Object doesn't exist
How to reproduce:

I will provide you with an E2 that is causing error (see files below, keep in mind that you need to download both of them since one of them is lib)

1. press on the button with gmod icon on top left
2. Close it with the button on the right side of the window.
3. Get the error
4. Error:

```
[wire-master] addons/wire-master/lua/entities/gmod_wire_egp/lib/egplib/objectcontrol.lua:182: attempt to index local 'obj' (a nil value)
  1. PerformReorder - addons/wire-master/lua/entities/gmod_wire_egp/lib/egplib/objectcontrol.lua:182
   2. Receive - addons/wire-master/lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua:551
    3. func - addons/wire-master/lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua:557
     4. unknown - lua/includes/extensions/net.lua:33

sv: Expression 2 (TurboOS  Version 0.0.1
Made By KemGus
): Error(s) occured while executing 'postexecute' hook:
addons/wire-master/lua/entities/gmod_wire_egp/lib/egplib/objectcontrol.lua:182: attempt to index local 'obj' (a nil value)
```
[egp_functions.txt](https://github.com/wiremod/wire/files/10889073/egp_functions.txt)
[os_.txt](https://github.com/wiremod/wire/files/10889074/os_.txt)

Notes:
For some reason, when I close the window by clicking on the icon again, it doesn't error that often.
Sometimes error won't kick in, so the issue seems a bit random.

### LINKED ISSUE #2553: wirelink:egpPos(n,v2) works incorrectly when used on egpLine()
Basicly when one tries to use egpPos(number index, vector2 pos) on an egpLine, it moves the starting end of the line instead of the whole thing

Link your egp v3 hud to your e2
Put wirelink:egpLine(number index, vector2 pos1, vector2 pos2) in your if first statement
Next put wirelink:egpPos(number index, vector2 pos) under it
upload and exit

**Expected behavior**
Expected line to move to said position

**Screenshots**

Code screenshot
code

Expected behavior screenshot
expected behavior

Result screenshot
result

### PR DESCRIPTION
**Pull request by Denneisk**
This is a very broad group of changes for EGPs, mostly aimed at clientsided performance improvements.

#### Additions and changes:
- EGP
  - *RenderTable_Indices* field added which indexes *RenderTabe*'s elements by their EGP index.[^1] Should help on larger EGPs. Moves the performance cost from finding to creation/removal/moving.
  - Transmissions culled by PVS. Certain transmissions still broadcasted to keep users in sync. Still needs testing and work.
  - Various optimizations across all things EGP.
  - E2: `egpRunOnQueue()` deprecated by event `egpQueueFinished()`
  - E2: `egpCreate()` and `egpManipulate()` for creating and manipulating EGP objects by raw table values.[^2] Performance is roughly the same.
  - E2: `wirelink:egpPos(n, v2, ...)` variadic egpPos for changing poly/line positions.
  - E2: display HUD users with `egpConnectedUsers()`.
- GPULib
  - ~~Change to GetRenderTargetEx, fixes forced trilinear filtering (now actually uses anisotropic).~~
  - ~~Combined finding an unused RT with creating one.~~
- Misc
  - ~~Tone mapping fixed for certain screens.[^3] Haven't gotten to applying fix to others. Causes washed out displays in some instances, but this is likely due to people not being able to see actual neutral EGP brightness.~~

#### Other fixes:
Fixes #2553 [(diff)](/wiremod/wire/pull/2571/files#diff-7cf3d1fa0bafe5a4caa12fa8cce7ea06f0f3913baef9b61832e38c377890e9bbR682), ~~fixes #2190 [(diff in usefulfunctions.lua)](/wiremod/wire/pull/2571/files#diff-fa68fa19dca64a922de121219b4b5384eba1a3217fa29e11ed2ba408ed3a1e5bR84)~~
[^1]: Fixes #2546 and #1476
[^2]: Fixes #1251 [(diff)](/wiremod/wire/pull/2571/files#diff-7cf3d1fa0bafe5a4caa12fa8cce7ea06f0f3913baef9b61832e38c377890e9bbR183)
[^3]: Fixes #2349 [(diff)](/wiremod/wire/pull/2571/files#diff-905c2a523b970d362ae1f3b138d9ad7bbce009e374c4a384f0cc8239307a74aaR74)

### CODE REVIEW

**Comment by wrefgtzweve** [comment]
If i remember correctly, EGP data is broadcasted to all players regardless if they're using it or not. It would be cool if it only sends EGP data once the chip is being used.

**Comment by Denneisk** [comment]
> If i remember correctly, EGP data is broadcasted to all players regardless if they're using it or not. It would be cool if it only sends EGP data to players using the chip.

EGP HUD data you mean? I believe that's already how it's done but I'll look into it.

**Comment by Divran** [comment]
> If i remember correctly, EGP data is broadcasted to all players regardless if they're using it or not. It would be cool if it only sends EGP data to players using the chip.

I don't remember how that worked but it sounds very likely given that when egp was made we only had usmg and if someone new walked up and started using the hud it would have taken ages to sync everything back then. But with the newer transfer method it should be a lot faster

**Comment by KemGus** [comment]
Finally somebody is fixing the issues in EGP!
I really hope to see further improvements in the future. Do you plan on making EGP great again?

File: lua/entities/gmod_wire_egp/lib/egplib/objectcontrol.lua [hunk 1]
```diff
@@ -179,7 +179,7 @@ function EGP:PerformReorder( Ent )
 	-- Second pass, remove objects from their original positions
 	for i=#Ent.RenderTable,1,-1 do
 		local obj = Ent.RenderTable[i]
-		if obj.ChangeOrder then
+		if obj and obj.ChangeOrder then
```
Comment (thegrb93) [inline lua/entities/gmod_wire_egp/lib/egplib/objectcontrol.lua:182]: This is just covering up a bug caused by something else. There should not be a nil partway in the RenderTable.

File: lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua
Comment (thegrb93) [inline lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua:1]: Don't mix format/style changes with code changes. It makes reviews more painful than they need to be.

**Review by thegrb93 — CHANGES_REQUESTED** [review]

File: lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua
Comment (Denneisk) [reply lua/entities/gmod_wire_egp/lib/egplib/transmitreceive.lua:1]: Yeah, sorry about that.

**Comment by Denneisk** [comment]
Here's your proper solution lol

File: lua/entities/gmod_wire_egp/init.lua [hunk 2]
```diff
@@ -48,4 +49,4 @@ function ENT:GetEGPOwner()
 	return false
 end
 
-function ENT:UpdateTransmitState() return TRANSMIT_ALWAYS end
+function ENT:UpdateTransmitState() return TRANSMIT_PVS end
```
Comment (thegrb93) [inline lua/entities/gmod_wire_egp/init.lua:52]: This shouldn't be changed. The entity should always exist for clients in order to stay in sync.

File: lua/entities/gmod_wire_expression2/core/egpfunctions.lua [hunk 3]
```diff
@@ -180,6 +180,83 @@ e2function void wirelink:egpTextLayout( number index, string text, vector2 pos,
...
+	
+	if args.parentEntity and types.parentEntity ~= "e" then self:throw("Argument parentEntity is not an entity!") end
+	
+	return true
+end
+
+-- Creates an object of name with all arguments listed in args
+e2function void wirelink:egpCreate(string objectName, table args)
+	if not EGP:IsAllowed(self, this) then return end
+	if not EGP.Objects.Names[objectName] then self:throw("EGP object name '" .. objectName .. "' does not exist!") end
```
Comment (thegrb93) [inline lua/entities/gmod_wire_expression2/core/egpfunctions.lua:234]: Need to make sure you're doing `return self:throw(`

**Comment by Denneisk** [comment]
A quick look at some EGP creations doesn't reveal anything besides the expected; E2s that use DrawTopLeft will have to be modified in lieu of the fix.

**Comment by Denneisk** [comment]
Putting back in drafts to address some recent considerations.

**Comment by thegrb93** [comment]
EGP never failed to destroy the willpower of anyone who touches it

**Pull request closed without merging**